src/primitives/compact_action.rs

```diff
@@ -0,0 +1,181 @@
+//! Defines actions for Orchard shielded outputs and compact action for light clients.
+
+// Review hint: this file is largely derived from src/note_encryption.rs
+
+use core::fmt;
+
+use zcash_note_encryption::{note_bytes::NoteBytes, EphemeralKeyBytes, ShieldedOutput};
+
+use crate::{
+    action::Action,
+    note::{ExtractedNoteCommitment, Nullifier, Rho},
+};
+
+use super::{orchard_domain::OrchardDomain, orchard_primitives::OrchardPrimitives};
+
+impl<A, P: OrchardPrimitives> ShieldedOutput<OrchardDomain<P>> for Action<A, P> {
+    fn ephemeral_key(&self) -> EphemeralKeyBytes {
+        EphemeralKeyBytes(self.encrypted_note().epk_bytes)
+    }
+
+    fn cmstar(&self) -> &ExtractedNoteCommitment {
+        self.cmx()
+    }
+
+    fn cmstar_bytes(&self) -> [u8; 32] {
+        self.cmx().to_bytes()
+    }
+
+    fn enc_ciphertext(&self) -> Option<&P::NoteCiphertextBytes> {
+        Some(&self.encrypted_note().enc_ciphertext)
+    }
+
+    fn enc_ciphertext_compact(&self) -> P::CompactNoteCiphertextBytes {
+        P::CompactNoteCiphertextBytes::from_slice(
+            &self.encrypted_note().enc_ciphertext.as_ref()[..P::COMPACT_NOTE_SIZE],
+        )
+        .unwrap()
```

Comment:
_From str4d ([r2599565711](https://github.com/zcash/orchard/pull/471#discussion_r2599565711))_

```suggestion
        .expect("P::CompactNoteCiphertextBytes should have size P::COMPACT_NOTE_SIZE")
```

Reply:
FIXED